### ISSUE TITLE
Segfault with xmlCopyPropList

### ISSUE THREAD

**Issue by gwillem** [facebook/hhvm#4682]
I get reproducible segfaults on multiple sites/servers running Magento (all Ubuntu 12.04 64bit with all security updates applied):
```
Host: xxxx
ProcessID: 22911
ThreadID: 7f8802bff700
ThreadPID: 22927
Name: unknown program
Type: Segmentation fault
Runtime: hhvm
Version: tags/HHVM-3.5.0-0-ga90f4733cfa0d8fefdafc7433f758f78cdc75424
DebuggerCount: 0

ThreadType: Web Request
Server_SERVER_NAME: www.xxxxx.nl
Server: www.xxxxx.nl
URL: /ajaxcart/ajax/getcart

# 0  xmlCopyPropList at /usr/lib/x86_64-linux-gnu/libxml2.so.2:0
# 1  HPHP::c_DOMDocument_ni_importNode(HPHP::ObjectData*, HPHP::Object const&, bool) at /usr/bin/hhvm:0
# 2  void HPHP::Native::callFunc<false, false>(HPHP::Func const*, void*, HPHP::TypedValue*, HPHP::TypedValue&) at /usr/bin/hhvm:0
# 3  HPHP::TypedValue* HPHP::Native::methodWrapper<false, false>(HPHP::ActRec*) at /usr/bin/hhvm:0

```
Tested with the following:
```
Version: tags/HHVM-3.3.0-0-g0a3cfb87b8a353fc7e1d15374f4adc413e37aba9
Version: tags/HHVM-3.3.1-0-ge0c98e21167b425dddf1fc9efe78c9f7a36db268
Version: tags/HHVM-3.4.2-0-g40d85d7386b3342ab848fc45c6892433c89ef5b2
Version: tags/HHVM-3.5.0-0-ga90f4733cfa0d8fefdafc7433f758f78cdc75424
```

All using the latest ```libxml2 2.7.8.dfsg-5.1ubuntu4.11``` from Ubuntu Precise (from security update).

Strangely enough, I've been running hhvm-nightly (3.3.0-dev+2014.08.01) since its release __without__ these segfaults. The nightly version is between 3.3.0 and 3.3.1, which doesn't make sense. For now, I've downgraded all my machines to this nightly version. 

Will try to reproduce with -dbg version if I find some time.

**Comment by gwillem** [comment]
Tried with latest hhvm-nightly-dbg (version 2015.01.07), and got this, but it seems unrelated:

```
ProcessID: 25160
ThreadID: 7f2c06fff700
ThreadPID: 25165
Name: unknown program
Type: Aborted
Runtime: hhvm
Version: heads/master-0-g263ddf7c5ada970a59d4814ef5c69f4f6d57b08b
DebuggerCount: 0

ThreadType: Web Request
Server_SERVER_NAME: www.xxxx.nl
Server: www.xxx.nl
URL: /xxxx.html?q=4176

# 0  raise at /lib/x86_64-linux-gnu/libc.so.6:0
# 1  abort at /lib/x86_64-linux-gnu/libc.so.6:0
# 2  __uselocale at /lib/x86_64-linux-gnu/libc.so.6:0
# 3  __assert_fail at /lib/x86_64-linux-gnu/libc.so.6:0
# 4  HPHP::jit::(anonymous namespace)::isVMFrame(HPHP::ExecutionContext const*, HPHP::ActRec const*) at fixup.cpp:0
# 5  HPHP::jit::FixupMap::fixupWork(HPHP::ExecutionContext*, HPHP::ActRec*) const at /usr/bin/hhvm:0
# 6  HPHP::jit::FixupMap::fixup(HPHP::ExecutionContext*) const at /usr/bin/hhvm:0
# 7  HPHP::jit::MCGenerator::syncWork() at /usr/bin/hhvm:0
# 8  HPHP::VMRegAnchor::VMRegAnchor() at /usr/bin/hhvm:0
# 9  HPHP::createBacktrace(HPHP::BacktraceArgs const&) at /usr/bin/hhvm:0
# 10 HPHP::debug_string_backtrace(bool, bool, long) at /usr/bin/hhvm:0
# 11 xmlCopyPropList at /usr/lib/x86_64-linux-gnu/libxml2.so.2:0
# 12 HPHP::c_DOMDocument_ni_importNode(HPHP::ObjectData*, HPHP::Object const&, bool) at /usr/bin/hhvm:0
# 13 HPHP::Native::callFuncInt64Impl(HPHP::TypedValue* (*)(HPHP::ActRec*), long*, int, double*, int) at /usr/bin/hhvm:0
# 14 void HPHP::Native::callFunc<false, false>(HPHP::Func const*, void*, HPHP::TypedValue*, HPHP::TypedValue&) at /usr/bin/hhvm:0
# 15 HPHP::TypedValue* HPHP::Native::methodWrapper<false, false>(HPHP::ActRec*) at /usr/bin/hhvm:0
```

**Comment by gwillem** [comment]
So to summarize, the only HHVM package that is working throughout all our installations is:
```
$ hhvm --version
HipHop VM 3.3.0-dev+2014.08.01 (rel)
Compiler: heads/master-0-g414e375f658717c0db5f3c7c0b2ca268e1b87373
Repo schema: 519c08c0b6b9e9eade7c65bc94461c08ef8b3e0c
Extension API: 20140702
```

**Comment by paulbiss** [comment]
Hmm, importNode is buggy in 3.3 and 3.6. There's a fix that went into 3.7 (it's also in the nightlies now) but we haven't decided if it's too large to pick into 3.6 (there's pretty much no chance it's going into 3.3). Have you tried this with 3.7 or a nightly?

**Comment by gwillem** [comment]
Can't test easily, as I was only able to reproduce it on production sites with live traffic. At the moment I don't have permission of the site owners to do more experimentation. Too bad I don't have an isolated test case.

**Comment by paulbiss** [comment]
My guess is that this has been fixed in 3.7+, but back-porting the fix would be non-trivial.

**Comment by paulbiss** [comment]
Closing this as there was a major rewrite of the libxml extensions in 3.7 and we've not seen new reports since. If anyone finds an isolated repro please comment here.
